### PR TITLE
Remotely render contributions epic (A/B test)

### PR DESCRIPTION
**Pull request by nicl**
## What does this change?

Adds an A/B test to compare a new remotely-rendered epic against the current default. The idea is to run this as a 10% test for enough time to confirm the two perform equivalently across the key metrics.

**UPDATE:** the following Ophan events happen at the moment:

```
{"Query string":{"viewId":"k6ga1y0avh6xkqxrdcm0","componentEvent":"{\"component\":{\"componentType\":\"ACQUISITIONS_EPIC\",\"products\":[\"CONTRIBUTION\",\"MEMBERSHIP_SUPPORTER\"],\"campaignCode\":\"gdnwb_copts_memco_remote_epic_test_remote\",\"id\":\"gdnwb_copts_memco_remote_epic_test_remote\"},\"abTest\":{\"name\":\"RemoteRenderEpic\",\"variant\":\"remote\"},\"action\":\"INSERT\"}"}}

{"Query string":{"viewId":"k6ga1y0avh6xkqxrdcm0","abTestRegister":"{\"RemoteRenderEpic\":{\"variantName\":\"remote\",\"complete\":\"false\",\"campaignCodes\":[\"gdnwb_copts_memco_remote_epic_test_remote\"]}}"}}

{"Query string":{"viewId":"k6ga1y0avh6xkqxrdcm0","componentEvent":"{\"component\":{\"componentType\":\"ACQUISITIONS_EPIC\",\"products\":[\"CONTRIBUTION\",\"MEMBERSHIP_SUPPORTER\"],\"campaignCode\":\"gdnwb_copts_memco_remote_epic_test_remote\",\"id\":\"gdnwb_copts_memco_remote_epic_test_remote\"},\"abTest\":{\"name\":\"RemoteRenderEpic\",\"variant\":\"remote\"},\"action\":\"VIEW\"}"}}

{"Query string":{"viewId":"k6ga1y0avh6xkqxrdcm0","abTestRegister":"{\"RemoteRenderEpic\":{\"variantName\":\"remote\",\"complete\":\"true\",\"campaignCodes\":[\"gdnwb_copts_memco_remote_epic_test_remote\"]}}"}} 
```
## Does this change need to be reproduced in dotcom-rendering ?

- [ ] No
- [x] Yes (please indicate your plans for DCR Implementation)

(We are going to do a subsequent A/B test on DCR and are planning this with the DCR team.)

## Screenshots

control:
Screenshot 2020-02-10 at 13 38 47

remote variant:
Screenshot 2020-02-10 at 13 38 47

## What is the value of this and can you measure success?

(Part of Slot Machine project.)

This is part of work to:

* support contributions on DCR and cross-platform generally
* decouple contributions (testing, changing, deploying, etc.) from the Frontend codebase
* faster development (shorter feedback cycles and better tech choices for css etc.)
* inform and progress the wider Slot Machine project

### Accessibility test checklist

**NIC: TODO BEFORE RELEASE.**

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: common/app/conf/switches/ABTestSwitches.scala [hunk 1]
```diff
@@ -115,4 +115,14 @@ trait ABTestSwitches {
...
     exposeClientSide = true
   )
+
+  Switch(
+    ABTests,
+    "ab-remote-render-epic",
+    "A/B test local vs remote render of default epic, to validate Slot Machine approach and work to date",
+    owners = Seq(Owner.withGithub("tjmw"), Owner.withGithub("nicl")),
+    safeState = On,
+    sellByDate = new LocalDate(2020, 2, 4),
```
Comment (tjmw) [inline common/app/conf/switches/ABTestSwitches.scala]: Reminder that we need to update this date 📆

File: static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js [hunk 2]
```diff
@@ -240,6 +240,74 @@ const articleViewCountIsOk = (
...
+    element: HTMLElement,
+    parentTest: EpicABTest,
+    campaignCode: ?string,
+    trackingCampaignId: string,
+    products: $ReadOnlyArray<OphanProduct>,
+    showTicker: boolean = false
+) => {
+    // top offset of 18 ensures view only counts when half of element is on screen
+    const inView = elementInView(element, window, {
+        top: 18,
```
Comment (tjmw) [inline static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: Out of interest, where does the `18` here come from? I wonder if it's worth giving that value a name to clarify.

Comment (tjmw) [reply static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: Oh, I just noticed this was moved from further down (so not really introduced here).

File: static/src/javascripts/projects/common/modules/experiments/ab.js [hunk 3]
```diff
@@ -52,7 +54,9 @@ export const getEpicTestToRun = memoize(
                     test => !test.highPriority
                 );
 
+                // TODO prioritise our hardcoded test at the moment
```
Comment (tjmw) [inline static/src/javascripts/projects/common/modules/experiments/ab.js]: Reminder for us to remove (or update) this comment.

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+                            if (!parent) {
+                                return;
+                            }
+
+                            const container = document.createElement('div');
+                            parent.insertBefore(container, target);
+
+                            // use Shadow Dom if found
+                            if (container.attachShadow) {
```
Comment (tjmw) [inline static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: Given that we're now explicitly checking for availability of the Shadow DOM API (and falling back if not) I wonder if we should remove the polyfill. My experience in IE11 was that the polyfill makes `.attackShadow` available, but doesn't provide the isolation you'd expect (page styles were still bleeding into the epic).

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+const products = ['CONTRIBUTION', 'MEMBERSHIP_SUPPORTER'];
+
+const remoteRenderTest = {
+    id: 'RemoteRenderEpic',
+    campaignId,
+
+    highPriority: true,
+
+    start: '2020-01-01',
+    expiry: '2020-02-29',
```
Comment (tjmw) [inline static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: Reminder for us to clarify/update these dates.

File: common/app/conf/switches/ABTestSwitches.scala (hunk 1 above)
Comment (nicl) [reply common/app/conf/switches/ABTestSwitches.scala]: thanks yeah! Maybe til the end of the month?

File: static/src/javascripts/projects/common/modules/experiments/ab.js [hunk 6]
```diff
@@ -53,6 +54,7 @@ export const getEpicTestToRun = memoize(
                 );
 
                 return firstRunnableTest<EpicABTest>([
+                    hardCodedPriorityEpicTest,
```
Comment (tomrf1) [inline static/src/javascripts/projects/common/modules/experiments/ab.js]: I think this name makes this part even more confusing. Perhaps just reference `remoteRenderEpic` directly here, as it's a temporary special case?

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+    author: 'Nicolas Long',
+    description:
+        'A/B test local vs remote render of default epic, to validate Slot Machine approach and work to date',
+    successMeasure: 'Conversion rate',
+    idealOutcome: 'No difference between control and variant',
+
+    audienceCriteria: 'All',
+    audience: 1,
```
Comment (tomrf1) [inline static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: to pick up 10% of the audience, set this to `0.1`

Comment (gtrufitt) [reply static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: go big or go home

File: static/src/javascripts/projects/common/modules/experiments/ab.js (hunk 6 above)
Comment (nicl) [reply static/src/javascripts/projects/common/modules/experiments/ab.js]: @tomrf1 yeah I agree. I only introduced this indirection to make the tests pass to be honest, but can explain IRL as would definitely prefer a better solution too!

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js (hunk 7 above)
Comment (nicl) [reply static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: ha thanks! 100% wouldn't have been so great 🤦‍♂

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+                const localisation = {
+                    countryCode: geolocation,
+                };
+
+                const targeting = {
+                    contentType: page.contentType,
+                    sectionName: page.sectionName,
+                    shouldHideReaderRevenue: page.shouldHideReaderRevenue,
+                    isMinuteArticle: page.isLiveBlog, // Is this the same thing?
```
Comment (tjmw) [inline static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: This isn't the same thing! Fix this.

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                    localisation,
+                    targeting,
+                });
+
+                const trackingCampaignId = `epic_${campaignId}`; // note exposed on ABTest unfortunately
+
+                emitBeginEvent(trackingCampaignId);
+                console.log('epic - beginning remote variant...');
+
+                setupClickHandling(
```
Comment (tjmw) [inline static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: Should we defer this until we've inserted the epic into the page?

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js (hunk 8 above)
Comment (tjmw) [reply static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: Fixed in cadd141

**Comment by tjmw** [comment]
Hi @tomrf1, after our discussion yesterday we updated some of the test configuration (see 494f05e). Rather than hardcoding the values we use for `campaignCode`, `abTestName` and `abTestVariant` we're now using values from the `EpicABTest` and `EpicVariant` yielded to the callback, as we discussed. In the short term I've commented the relevant lines with the actual string values we get:

```js
campaignCode: variant.campaignCode, // "gdnwb_copts_memco_remote_epic_test_remote"
abTestName: test.id, // "RemoteRenderEpic"
abTestVariant: variant.id, // "remote"
```

These look like they fit with values I can see being used elsewhere, but please let us know if anything doesn't look quite right.

These are the resulting URLs we build up for the epic in each case:

## Variant (remote)

https://support.theguardian.com/uk/contribute?REFPVID=k6c1h2zxbmpvctihqcsd&INTCMP=gdnwb_copts_memco_remote_epic_test_remote&acquisitionData=%7B%22source%22%3A%22GUARDIAN_WEB%22%2C%22componentId%22%3A%22gdnwb_copts_memco_remote_epic_test_remote%22%2C%22componentType%22%3A%22ACQUISITIONS_EPIC%22%2C%22campaignCode%22%3A%22gdnwb_copts_memco_remote_epic_test_remote%22%2C%22abTest%22%3A%7B%22name%22%3A%22RemoteRenderEpic%22%2C%22variant%22%3A%22remote%22%7D%2C%22referrerPageviewId%22%3A%22k6c1h2zxbmpvctihqcsd%22%2C%22referrerUrl%22%3A%22http%3A%2F%2Flocalhost%3A9000%2Fbusiness%2F2020%2Fjan%2F07%2Fnissan-vows-to-hold-carlos-ghosn-accountable-for-serious-misconduct%22%7D

Decoding the acquisition data gives:

```json
{
  "source": "GUARDIAN_WEB",
  "componentId": "gdnwb_copts_memco_remote_epic_test_remote",
  "componentType": "ACQUISITIONS_EPIC",
  "campaignCode": "gdnwb_copts_memco_remote_epic_test_remote",
  "abTest": {
    "name": "RemoteRenderEpic",
    "variant": "remote"
  },
  "referrerPageviewId": "k6c1h2zxbmpvctihqcsd",
  "referrerUrl": "http://localhost:9000/business/2020/jan/07/nissan-vows-to-hold-carlos-ghosn-accountable-for-serious-misconduct"
}
```

## Control (local)

https://support.theguardian.com/uk/contribute?REFPVID=k6c1k2e1cct41lzz8ms4&INTCMP=gdnwb_copts_memco_remote_epic_test_control&acquisitionData=%7B%22source%22%3A%22GUARDIAN_WEB%22%2C%22componentId%22%3A%22gdnwb_copts_memco_remote_epic_test_control%22%2C%22componentType%22%3A%22ACQUISITIONS_EPIC%22%2C%22campaignCode%22%3A%22gdnwb_copts_memco_remote_epic_test_control%22%2C%22abTest%22%3A%7B%22name%22%3A%22RemoteRenderEpic%22%2C%22variant%22%3A%22control%22%7D%2C%22referrerPageviewId%22%3A%22k6c1k2e1cct41lzz8ms4%22%2C%22referrerUrl%22%3A%22http%3A%2F%2Flocalhost%3A9000%2Fbusiness%2F2020%2Fjan%2F07%2Fnissan-vows-to-hold-carlos-ghosn-accountable-for-serious-misconduct%23ab-RemoteRenderEpic%3Dcontrol%22%7D

Decoding the acquisition data gives:

```json
{
  "source": "GUARDIAN_WEB",
  "componentId": "gdnwb_copts_memco_remote_epic_test_control",
  "componentType": "ACQUISITIONS_EPIC",
  "campaignCode": "gdnwb_copts_memco_remote_epic_test_control",
  "abTest": {
    "name": "RemoteRenderEpic",
    "variant": "control"
  },
  "referrerPageviewId": "k6c1k2e1cct41lzz8ms4",
  "referrerUrl": "http://localhost:9000/business/2020/jan/07/nissan-vows-to-hold-carlos-ghosn-accountable-for-serious-misconduct#ab-RemoteRenderEpic=control"
}
```

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    successMeasure: 'Conversion rate',
+    idealOutcome: 'No difference between control and variant',
+
+    audienceCriteria: 'All',
+    audience: 0.1,
+    audienceOffset: 0,
+
+    geolocation,
+
+    canRun: () => geolocation !== 'US',
```
Comment (tjmw) [inline static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: Looking at other examples, it seems that this is enough to exclude the US.

Comment (nicl) [reply static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: Nice! 👍

File: static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+                const payload = JSON.stringify({
+                    tracking,
+                    localisation,
+                    targeting,
+                });
+
+                const trackingCampaignId = `epic_${campaignId}`; // note exposed on ABTest unfortunately
+
+                emitBeginEvent(trackingCampaignId);
+                console.log('epic - beginning remote variant...');
```
Comment (tomrf1) [inline static/src/javascripts/projects/common/modules/experiments/tests/remote-render-epic.js]: remove log?

**Review by tomrf1 — APPROVED** [review]
great work!

**Pull request merged**